### PR TITLE
Add create-import-video-file-job command

### PR DESCRIPTION
**Pull request by fflorent**
This is a WIP but I request some feedback ;).

This PR is meant to provide a command to import a video file for a given video UUID. Here are some use case for such a command:
- You want to restore a corrupted video file ;
- You want to support another resolution for all your videos and the server don't have a powerful CPU → You realize that you'd better transcode (with the right options) on another machine, upload it to the server and import the video file (#560)  ;
- You uploaded a video with a low resolution and want to propose a higher one ;
- You'd like to update something on your video (fixing something wrong, give more details, etc) ;

On all these use case, you'd like to keep the meta-data on it (descriptions, number of views, comments, etc).

So I propose a new command:
```bash
$ node ./dist/scripts/create-import-video-file-job.js -v [videoUUID] -i [videoFile]
```

Where:
 - `videoUUID` is the UUID of an existing video
 - `videoFile` is the video which replaces an existing video file or add a new resolution for it 

If you think these use case are justified and correctly addressed, here are some TODO:

- [ ] write tests (yeah, I shamefully submitted without any of them whereas I usually do TDD);

And some questions:

- [x] Is `import-video-file` meaningful? Probably far from being perfect, suggestions welcome!
- [x] Is `video-file.ts` the right place to add the function `processVideoImport`? Not really sure;
- [x] Not sure of what should be done when the import is finished. I call the same callback for the transcoding job, but not really sure;

Nitpicking and bikeshedding are welcome! I am conscious lots of things should be improved and I'll take comments into account humbly! ;)

Cheers!
Florent

### CODE REVIEW

File: server/models/video/video.ts [hunk 1]
```diff
@@ -1315,6 +1316,29 @@ export class VideoModel extends Model<VideoModel> {
...
+      resolution: (await getVideoFileResolution(inputFilePath)).videoFileResolution,
+      extname: extname(inputFilePath),
+      size: (await statPromise(inputFilePath)).size,
+      videoId: this.id
+    })
+
+    const outputPath = this.getVideoFilePath(newVideoFile)
+    const oldVideoFile = this.VideoFiles.find(videoFile => videoFile.resolution === newVideoFile.resolution)
+    if (oldVideoFile) {
+      oldVideoFile.destroy()
```
[the remark this answers is not in the document]
Comment (fflorent) [reply server/models/video/video.ts]: OK, I think removing the oldVideoFile from the array should help. Also there is a momentary in which the video is not available (during the video copy).

File: server/initializers/constants.ts [hunk 2]
```diff
@@ -74,6 +74,7 @@ const JOB_ATTEMPTS: { [ id in JobType ]: number } = {
   'activitypub-http-unicast': 5,
   'activitypub-http-fetcher': 5,
   'activitypub-follow': 5,
+  'video-file-import': 5,
```
Comment (Chocobozzz) [inline server/initializers/constants.ts]: 1 like video-file

File: server/models/video/video.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Chocobozzz) [reply server/models/video/video.ts]: I think that just updating the size, infohash and extname should be better

**Comment by fflorent** [comment]
@Chocobozzz Thanks for your feedback. I'll write the tests this afternoon!

**Comment by fflorent** [comment]
@Chocobozzz Not sure what is the intent behind these steps:
> - Upload a video on server 2 (transcoded is enabled)
> - Import a vile
> - Check server 1 and 2 has this video with your imported file

**Comment by Chocobozzz** [comment]
> Not sure what is the intent behind these steps:

To test if **replacing** an existing video file works (ie the infohash and size are updated)

**Comment by fflorent** [comment]
Here is the work in progress on the tests. The second test doesn't pass for some reasons… https://github.com/Chocobozzz/PeerTube/pull/619/commits/4c998fef8736ec2b2887c85b554b28a5395846b7

I have two questions for now:
 - I am not sure what to pass to `overrideConfig`;
 - It seems like the extension is not updated in the magnet;

If anyone can give me some clue, that would be very appreciated :).
Thanks in advance,
Florent

File: server/tests/cli/create-import-video-file-job.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  ServerInfo,
+  setAccessTokensToServers,
+  uploadVideo,
+  wait,
+  getVideo, flushAndRunMultipleServers, doubleFollow
+} from '../utils'
+
+function assertVideoProperties (video: VideoFile, resolution: number, extname: string) {
+  expect(video).to.have.nested.property('resolution.id', resolution)
+  expect(video).to.have.property('magnetUri').that.matches(new RegExp(`\\.${extname}$`))
```
Comment (Chocobozzz) [inline server/tests/cli/create-import-video-file-job.ts]: Just `indexOf(str) !== -1` should be enought. Please test the `torrentUrl` and `fileUrl` too

Comment (Chocobozzz) [reply server/tests/cli/create-import-video-file-job.ts]: Please check the `size` too

File: server/tests/cli/create-import-video-file-job.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      const { data: videos } = (await getVideosList(server.url)).body
+      expect(videos).to.have.lengthOf(2)
+
+      let infoHashes: { [ id: number ]: string }
+
+      const video = videos.find(({ uuid }) => uuid = video2UUID)
+      const videoDetail: VideoDetails = (await getVideo(server.url, video.uuid)).body
+
+      console.log(videoDetail.files)
+      expect(videoDetail.files).to.have.lengthOf(2)
```
Comment (Chocobozzz) [inline server/tests/cli/create-import-video-file-job.ts]: server 2 transcodes so files should have length of 4

File: server/tests/cli/create-import-video-file-job.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      const [originalVideo, transcodedVideo] = videoDetail.files
+      assertVideoProperties(transcodedVideo, 320, 'webm')
```
Comment (Chocobozzz) [inline server/tests/cli/create-import-video-file-job.ts]: transcoded videos are always mp4

File: server/tests/utils/server/servers.ts [hunk 6]
```diff
@@ -35,7 +35,7 @@ interface ServerInfo {
   }
 }
 
-function flushAndRunMultipleServers (totalServers) {
+function flushAndRunMultipleServers (totalServers, configOverrideFn = (numServer: number) => undefined) {
```
Comment (Chocobozzz) [inline server/tests/utils/server/servers.ts]: please remove configOverrideFn

File: server/tests/utils/server/servers.ts [hunk 7]
```diff
@@ -53,7 +53,7 @@ function flushAndRunMultipleServers (totalServers) {
         for (let j = 1; j <= totalServers; j++) {
           // For the virtual buffer
           setTimeout(() => {
-            runServer(j).then(app => anotherServerDone(j, app))
+            runServer(j, configOverrideFn(j)).then(app => anotherServerDone(j, app))
```
Comment (Chocobozzz) [inline server/tests/utils/server/servers.ts]: Please remove configOverrideFn

File: server/models/video/video.ts [hunk 8]
```diff
@@ -1315,6 +1316,38 @@ export class VideoModel extends Model<VideoModel> {
...
+
+    const outputPath = this.getVideoFilePath(updatedVideoFile)
+    await copyFilePromise(inputFilePath, outputPath)
+
+    const currentVideoFile = this.VideoFiles.find(videoFile => videoFile.resolution === updatedVideoFile.resolution)
+    const isNewVideoFile = !currentVideoFile
+
+    if (!isNewVideoFile) {
+      if (currentVideoFile.extname !== updatedVideoFile.extname) {
+        await unlinkPromise(this.getVideoFilePath(currentVideoFile))
```
Comment (Chocobozzz) [inline server/models/video/video.ts]: `this.removeFile(currentVideoFile)`

File: server/models/video/video.ts [hunk 9]
```diff
@@ -1315,6 +1316,38 @@ export class VideoModel extends Model<VideoModel> {
...
+      }
+      currentVideoFile.set('size', updatedVideoFile.size)
+      updatedVideoFile = currentVideoFile
+    }
+
+    await this.createTorrentAndSetInfoHash(updatedVideoFile)
+
+    await updatedVideoFile.save()
+
+    if (isNewVideoFile) {
```
Comment (Chocobozzz) [inline server/models/video/video.ts]: You need to delete the old one if it's not a new file

Comment (fflorent) [reply server/models/video/video.ts]: Well, in this case, either:
 - the extension name has changed, so the file is removed (see above)
 - the extension name is the same, so is the whole file name. The old file is overridden by the `copyFilePromise` call

I think this cover all the cases :)

File: server/tests/utils/server/servers.ts (hunk 6 above)
Comment (fflorent) [reply server/tests/utils/server/servers.ts]: How do I disable transcoding then for the first server? (if I should)

**Comment by Chocobozzz** [comment]
You don't need to disable transcoding manually

See https://github.com/Chocobozzz/PeerTube/blob/develop/config/test-1.yaml#L31 and https://github.com/Chocobozzz/PeerTube/blob/develop/config/test-2.yaml#L28

**Comment by fflorent** [comment]
So server[0] is automatically mapped to the test-1.yaml config, and server[1] to the test-2.yaml?

Florent

**Comment by Chocobozzz** [comment]
Yes

**Pull request merged**